core/cppgc.rs

```diff
@@ -1,74 +1,88 @@
 // Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
 
+use crate::JsRuntime;
 use std::any::TypeId;
+pub use v8::cppgc::GarbageCollected;
 
-pub trait GcResource {}
+const CPPGC_TAG: u16 = 1;
 
-struct CppGcObject<T: GcResource> {
+#[repr(C)]
+struct CppGcObject<T: GarbageCollected> {
   tag: TypeId,
-  member: T,
+  member: v8::cppgc::Member<T>,
 }
 
-impl<T: GcResource> v8::cppgc::GarbageCollected for CppGcObject<T> {
-  fn trace(&self, _: &v8::cppgc::Visitor) {}
+impl<T: GarbageCollected> v8::cppgc::GarbageCollected for CppGcObject<T> {
+  fn trace(&self, visitor: &v8::cppgc::Visitor) {
+    visitor.trace(&self.member);
+  }
 }
 
-pub(crate) const DEFAULT_CPP_GC_EMBEDDER_ID: u16 = 0xde90;
-
-const EMBEDDER_ID_OFFSET: i32 = 0;
-const SLOT_OFFSET: i32 = 1;
-const FIELD_COUNT: usize = 2;
+pub fn make_cppgc_template<'s>(
+  scope: &mut v8::HandleScope<'s, ()>,
+) -> v8::Local<'s, v8::FunctionTemplate> {
+  fn empty(
+    _scope: &mut v8::HandleScope,
+    _args: v8::FunctionCallbackArguments,
+    _rv: v8::ReturnValue,
+  ) {
+  }
+  v8::FunctionTemplate::new(scope, empty)
+}
 
-pub fn make_cppgc_object<'a, T: GcResource + 'static>(
+pub fn make_cppgc_object<'a, T: GarbageCollected + 'static>(
   scope: &mut v8::HandleScope<'a>,
   t: T,
 ) -> v8::Local<'a, v8::Object> {
-  let templ = v8::ObjectTemplate::new(scope);
-  templ.set_internal_field_count(FIELD_COUNT);
+  let state = JsRuntime::state_from(scope);
+  let templ =
+    v8::Local::new(scope, state.cppgc_template.borrow().as_ref().unwrap());
+  let func = templ.get_function(scope).unwrap();
+  let obj = func.new_instance(scope, &[]).unwrap();
+
+  let heap = scope.get_cpp_heap().unwrap();
 
-  let obj = templ.new_instance(scope).unwrap();
-  let member = v8::cppgc::make_garbage_collected(
-    scope.get_cpp_heap().unwrap(),
-    Box::new(CppGcObject {
-      tag: TypeId::of::<T>(),
-      member: t,
-    }),
-  );
+  let member = unsafe {
+    v8::cppgc::make_garbage_collected(
+      heap,
+      CppGcObject {
+        tag: TypeId::of::<T>(),
+        member: v8::cppgc::make_garbage_collected(heap, t),
+      },
+    )
+  };
+
+  unsafe {
+    v8::Object::wrap::<CPPGC_TAG, CppGcObject<T>>(scope, obj, &member);
+  }
 
-  obj.set_aligned_pointer_in_internal_field(
-    EMBEDDER_ID_OFFSET,
-    &DEFAULT_CPP_GC_EMBEDDER_ID as *const u16 as _,
-  );
-  obj.set_aligned_pointer_in_internal_field(SLOT_OFFSET, member.handle as _);
   obj
 }
 
 #[allow(clippy::needless_lifetimes)]
-pub fn try_unwrap_cppgc_object<'sc, T: GcResource + 'static>(
+pub fn try_unwrap_cppgc_object<'sc, T: GarbageCollected + 'static>(
+  scope: &mut v8::HandleScope<'sc>,
   val: v8::Local<'sc, v8::Value>,
-) -> Option<&'sc T> {
+) -> v8::cppgc::Member<T> {
```

Comment:
If `try_unwrap_cppgc_object` returns a `Member` pointer then it needs to be `unsafe`, because the caller must not hold the pointer past the object's lifetime.

Reply:
I think previous return type was more user-friendly

Reply:
I would like to explore a replacement for make/unwrap using Member<T> in the future. `Option<&T>` (or more realistically `Option<cppgc::Ptr<'_, T>>`) would be nice, but its not actually possible in rust to get the correct lifetime on it at the moment (it can always be promoted off the stack), so it would still need to be an unsafe mess.